### PR TITLE
 Adjust name in resource;Reorganize and optimize log output

### PR DESCRIPTION
**Pull request by xiaomuqiao**
1、	change code name about resource and fix the issue that the outtime event is not finished
2、 wait for finishing send the log to kafka through cyclone-server
3、	Organize and 	Optimize log output

### CODE REVIEW

File: event/manager.go [hunk 1]
```diff
@@ -233,16 +237,15 @@ func (el *List) loadListFromEtcd(etcd *etcd.Client) {
 		event, err := loadEventFromJSON(jsonEvent)
 		if err != nil {
 			log.Errorf("load event from etcd err: %v", err)
-			return
+			continue
 		}
 		log.Infof("load event to list: %s", event.EventID)
 		w, err := LoadWorker(&event)
-		if err != nil {
```
Comment (gaocegege) [inline event/manager.go]: I don't know if I misunderstand you. No error handler here?

File: api/rest/log.go [hunk 2]
```diff
@@ -46,7 +47,15 @@ func getVersionLog(request *restful.Request, response *restful.Response) {
 		log.ErrorWithFields(message, log.Fields{"user_id": userID, "error": err})
 		getResponse.ErrorMessage = message
 	} else {
-		getResponse.Logs = result.Logs
+		arrary := strings.Split(result.Logs, "\n")
```
Comment (gaocegege) [inline api/rest/log.go]: s/arrary/array

File: event/manager.go [hunk 3]
```diff
@@ -354,7 +357,7 @@ func handlePendingEvents() {
 		event := *pendingEvents.GetFront()
 		err := handleEvent(&event)
 		if err != nil {
-			if err == Err_Unable_Support {
+			if err == resource.Err_Unable_Support {
```
Comment (gaocegege) [inline event/manager.go]: The name of this error `Err_Unable_Support` is not appropriate. Maybe we could use ErrUnableSupport or something else?

File: event/worker.go [hunk 4]
```diff
@@ -63,9 +62,7 @@ const (
 )
 
 var (
-	Err_Worker_Busy    = errors.New("Get worker docker host busy")
-	Err_Unable_Support = errors.New("Unable to support the request resource")
-	resourceManager    = resource.NewManager()
+	Err_Worker_Busy = errors.New("Get worker docker host busy")
```
Comment (gaocegege) [inline event/worker.go]: The names of these errors are not golang style :)

File: resource/resource.go [hunk 5]
```diff
@@ -37,18 +38,19 @@ type ResourceManager struct {
 }
 
 var (
-	resourceManager *ResourceManager
+	Err_Unable_Support = errors.New("Unable to support the request resource")
```
Comment (gaocegege) [inline resource/resource.go]: same here

File: websocket/application.go [hunk 6]
```diff
@@ -174,15 +174,37 @@ func PushTopic(wss *WSSession, pWatchLog *WatchLogPacket) {
...
-		byrLog := PacketPushLog(pWatchLog.Api, pWatchLog.UserId,
-			pWatchLog.ServiceId, pWatchLog.VersionId, string(msg.Value),
-			uuid.NewV4().String())
-		dpPacket := &DataPacket{
-			byrFrame:  byrLog,
-			nFrameLen: len(byrLog),
-			sSendTo:   wss.sSessionID,
+		str := string(msg.Value)
+
+		arrary := strings.Split(str, "\n")
```
Comment (gaocegege) [inline websocket/application.go]: s/arrary/array

File: websocket/application.go [hunk 7]
```diff
@@ -174,15 +174,37 @@ func PushTopic(wss *WSSession, pWatchLog *WatchLogPacket) {
...
+		for _, arr := range arrary {
```
Comment (gaocegege) [inline websocket/application.go]: Could we add some comments here, it's a little complex.

File: api/rest/init.go [hunk 8]
```diff
@@ -28,7 +28,7 @@ import (
 var (
 	// remoteManager is remote api manager.
 	remoteManager   *remote.Manager
-	resourceManager *resource.ResourceManager
+	resourceManager *resource.Manager
```
Comment (superxi911) [inline api/rest/init.go]: The old code should not be compiled successfully😓

File: api/rest/log.go [hunk 9]
```diff
@@ -46,7 +47,15 @@ func getVersionLog(request *restful.Request, response *restful.Response) {
...
+		for _, arr := range arrary {
+			if arr != "\r" && arr != "" {
+				if strings.HasPrefix(arr, "layer") != true {
```
Comment (superxi911) [inline api/rest/log.go]: const

Comment (superxi911) [inline api/rest/log.go]: need to wrap into a function 
```
func isPullImageLog(log string) bool {
    return strings.HasPrefix(arr, PULL_IMAGE_LOG_FLAG)
}
```

File: event/event.go [hunk 10]
```diff
@@ -123,7 +123,22 @@ func createServiceHandler(event *api.Event) error {
 		return err
 	}
 
-	return w.DoWork(event)
+	err = w.DoWork(event)
+	if err != nil {
+		return err
+	}
+
+	if event.Service.Repository.Webhook == api.GITHUB {
```
Comment (superxi911) [inline event/event.go]: Move to createVersionHandler

File: worker/ci/runner/docker.go [hunk 11]
```diff
@@ -240,13 +240,16 @@ func toBuildContainerConfig(dn *parser.DockerNode, b *Build, nodetype parser.Nod
 
 // start a container with the given CreateContainerOptions.
 func start(b *Build, cco *docker_client.CreateContainerOptions) (*docker_client.Container, error) {
-	// TODO: Inspect the image first, then decide whether pull the image or not.
-	log.InfoWithFields("About to pull the image.", log.Fields{"image": cco.Config.Image})
-	err := b.dockerManager.PullImage(cco.Config.Image)
-	if err != nil {
-		return nil, err
+	log.InfoWithFields("About to inspect the image.", log.Fields{"image": cco.Config.Image})
+	result, _ := b.dockerManager.IsImagePresent(cco.Config.Image)
```
Comment (gaocegege) [inline worker/ci/runner/docker.go]: maybe the error should be checked?

File: remote/provider/github.go [hunk 12]
```diff
@@ -86,7 +86,7 @@ func (g *GitHub) Authcallback(code, state string) (string, error) {
 
 	// Caicloud web address,eg caicloud.io
 	uiPath := osutil.GetStringEnv("CONSOLE_WEB_ENDPOINT", "http://localhost:8000")
-	redirectURL := fmt.Sprintf("%s/cyclone/add?type=github&code=%s&state=%s", uiPath, code, state)
+	redirectURL := fmt.Sprintf("%s/circle/add?type=github&code=%s&state=%s", uiPath, code, state)
```
Comment (superxi911) [inline remote/provider/github.go]: Should use cyclone

File: worker/cyclone-worker.go [hunk 13]
```diff
@@ -170,11 +172,18 @@ func createVersion(vcsManager *vcs.Manager, event *api.Event) {
 		output.Close()
 		event.Output = nil
 		worker_log.SetWatchLogFileSwitch(output.Name(), false)
+		// wait the send the log to kafka throuth cyclone server totally
```
Comment (gaocegege) [inline worker/cyclone-worker.go]: // Wait until log has been pushed into cyclone server.
By the way, why not use waitutil to do this. 🤔

File: remote/provider/gitlab.go [hunk 14]
```diff
@@ -87,7 +87,7 @@ func (g *GitLab) Authcallback(code, state string) (string, error) {
 
 	//caicloud web address,eg caicloud.io
 	uiPath := osutil.GetStringEnv("CONSOLE_WEB_ENDPOINT", "http://localhost:8000")
-	redirectURL := fmt.Sprintf("%s/cyclone/add?type=gitlab&code=%s&state=%s", uiPath, code, state)
+	redirectURL := fmt.Sprintf("%s/circle/add?type=gitlab&code=%s&state=%s", uiPath, code, state)
```
Comment (superxi911) [inline remote/provider/gitlab.go]: Should use cyclone

File: websocket/application.go [hunk 15]
```diff
@@ -174,15 +174,37 @@ func PushTopic(wss *WSSession, pWatchLog *WatchLogPacket) {
...
+			if arr != "\r" && arr != "" {
+				if strings.HasPrefix(arr, "layer") {
```
Comment (superxi911) [inline websocket/application.go]: wrap

File: api/rest/log.go (hunk 9 above)
Comment (superxi911) [inline api/rest/log.go]: We should do these before the log write into DB

**Comment by ddysher** [comment]
PR title doesn't make any sense to me 😆

File: api/rest/log.go (hunk 9 above)
Comment (xiaomuqiao) [inline api/rest/log.go]: yes, you are right

File: event/event.go (hunk 10 above)
Comment (xiaomuqiao) [inline event/event.go]: ok, this is my mistake.

File: remote/provider/github.go (hunk 12 above)
Comment (xiaomuqiao) [inline remote/provider/github.go]: ok, i cancel this modify.

File: remote/provider/gitlab.go (hunk 14 above)
Comment (xiaomuqiao) [inline remote/provider/gitlab.go]: the same as above

File: event/manager.go (hunk 1 above)
Comment (xiaomuqiao) [inline event/manager.go]: yes, i will delete "w, err := LoadWorker(&event) ". Here ie not need to do "LoadWorker".

File: api/rest/log.go (hunk 2 above)
Comment (xiaomuqiao) [inline api/rest/log.go]: e.. a terrible spelling

File: websocket/application.go (hunk 7 above)
Comment (xiaomuqiao) [inline websocket/application.go]: I will ..

File: worker/ci/runner/docker.go (hunk 11 above)
Comment (xiaomuqiao) [inline worker/ci/runner/docker.go]: yes, I will add.

File: worker/cyclone-worker.go (hunk 13 above)
Comment (xiaomuqiao) [inline worker/cyclone-worker.go]: It's my terrible english grammar.

File: event/manager.go (hunk 3 above)
Comment (xiaomuqiao) [inline event/manager.go]: Maybe i could remove the  '_' .

**Comment by xiaomuqiao** [comment]
@ddysher  I motify to use a little better title

**Comment by xiaomuqiao** [comment]
@superxi911  @gaocegege  PTAL

File: websocket/application.go [hunk 16]
```diff
@@ -174,16 +178,45 @@ func PushTopic(wss *WSSession, pWatchLog *WatchLogPacket) {
...
+				}
+			}
 		}
-		wss.Send(dpPacket)
 		time.Sleep(time.Millisecond * 100)
 	}
 	log.Infof("stop push %s to %s", sTopic, wss.GetSessionID())
 }
+
+func isDockerImageOperationLog(log string) bool {
```
Comment (superxi911) [inline websocket/application.go]: comments

File: worker/helper/output.go [hunk 17]
```diff
@@ -93,3 +105,7 @@ func getLogFromOutputFile(event *api.Event) (string, error) {
 	}
 	return string(content), nil
 }
+
+func isDockerImageOperationLog(log string) bool {
```
Comment (superxi911) [inline worker/helper/output.go]: comments

File: api/rest/log.go [hunk 18]
```diff
@@ -46,7 +47,15 @@ func getVersionLog(request *restful.Request, response *restful.Response) {
...
+					getResponse.Logs += arr
+					getResponse.Logs += "\n"
+				}
+			}
+		}
```
Comment (superxi911) [inline api/rest/log.go]: These codes can be deleted

File: websocket/application.go [hunk 19]
```diff
@@ -217,6 +217,7 @@ func PushTopic(wss *WSSession, pWatchLog *WatchLogPacket) {
 	log.Infof("stop push %s to %s", sTopic, wss.GetSessionID())
 }
 
+// isDockerImageOperationLog func that check the log whether is docker image' operating log.
```
Comment (superxi911) [inline websocket/application.go]: // isDockerImageOperationLog check the log whether is the log of pulling or pushing docker image.

**Comment by superxi911** [comment]
LGTM

**Pull request merged**